Fixes #11679 - Jetty 12.0.8 seems to leak connection when it encounters earlyEOF.

sbordet opened the pull request:
Changed HttpConnection.RequestHandler.earlyEOF() to produce EofException instead of BadMessageException, as it is more appropriate.

Changed handling of HttpChannelState.onFailure() to not fail the write side unless there is a pending write callback.

sbordet committed:
Fixes #11679 - Jetty 12.0.8 seems to leak connection when it encounters earlyEOF.

Changed HttpConnection.RequestHandler.earlyEOF() to produce EofException instead of BadMessageException, as it is more appropriate.

Changed handling of HttpChannelState.onFailure() to not fail the write side unless there is a pending write callback.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (changes requested):
There are plenty of tests expecting a response status 400 but now getting 500 instead.

sbordet committed:
Fixed checkstyle.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1120,12 +1120,12 @@ private boolean lockedIsWriting()
             Callback writeCallback = _writeCallback;
             _writeCallback = null;
-            if (writeCallback != null || fatal)
+            if (writeCallback != null)
```
This if has opposite condition of the following if, so that can be an else.

sbordet replied:
Good point. Simplified with an early return is `writeCallback == null`.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1120,12 +1120,12 @@ private boolean lockedIsWriting()
         }
 
-        private Runnable lockedFailWrite(Throwable x, boolean fatal)
+        private Runnable lockedFailWrite(Throwable x)
```
maybe rename to `lockedFailIfWritePending`?

sbordet replied:
I'd leave it with the current name.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1057,7 +1057,7 @@ public void earlyEOF()
             if (stream != null)
             {
-                BadMessageException bad = new BadMessageException("Early EOF");
+                EofException bad = new EofException("Early EOF");
```
I don't like that this sends a 500.    I accept that perhaps the request is not bad and it is a connection problem, but it is certainly not a server problem as the 500 suggests.
 How about:
```suggestion
                BadMessageException bad = new BadMessageException(null, new EofException("Early EOF"));
```

sbordet replied:
I think it should be a type of EOF exception because that's what it is.
I have introduced an `HttpEofException` so it is converted to a 400, but it is-a `IOException`, and it's quiet because it's a client close, so won't clutter the logs.

sbordet committed:
Early EOF events now produce a EofException that is also an HttpException.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (changes requested):
Still fails some related tests...

sbordet committed:
Fixed test that was failing due to change in behavior of the implementation.

Now failures only impact pending writes, so that it would be possible to write an HTTP error response.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1057,7 +1057,7 @@ public void earlyEOF()
             if (stream != null)
             {
-                BadMessageException bad = new BadMessageException("Early EOF");
+                HttpEofException bad = new HttpEofException("Early EOF");
```
Either hardcode the `"Early EOF"` message in the exception's `getReason()` method, or return the message passed to the constructor, but let's not mix both.

lorban reviewed (dismissed):
LGTM, besides one niggle.

sbordet committed:
Update after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw reviewed (dismissed):
I agree with @lorban comment... other than that LGTM.

joakime committed:
Applying changes from review
